cmd/storage-class.go

```diff
@@ -0,0 +1,213 @@
+
+	// Standard storage class implies more parity than Reduced redundancy storage class. So, Standard storage parity disks should be
+	// - greater than or equal to 2, if RRS parity is not set.
+	// - greater than RRS Parity, if RRS parity is set.
+	switch rrsParity {
+	case 0:
+		if ssParity < minimumParityDisks {
+			return errors.New("Standard storage class parity disks should be greater than or equal to " + strconv.Itoa(minimumParityDisks))
+		}
+	default:
+		if ssParity <= rrsParity {
+			return errors.New("Standard storage class parity disks should be greater than " + strconv.Itoa(rrsParity))
+		}
+	}
+
+	// Standard storage class parity should be less than or equal to N/2
+	if ssParity > len(globalEndpoints)/2 {
+		return errors.New("Standard storage class parity disks should be less than or equal to " + strconv.Itoa(len(globalEndpoints)/2))
+	}
+
+	return nil
+}
+
+// Returns the data and parity drive count based on storage class
+// If storage class is set using the env vars MINIO_STORAGE_CLASS_RRS and MINIO_STORAGE_CLASS_STANDARD
+// -- corresponding values are returned
+// If storage class is not set using environment variables, default values are returned
+// -- Default for Reduced Redundancy Storage class is, parity = 2 and data = N-Parity
+// -- Default for Standard Storage class is, parity = N/2, data = N/2
+// If storage class is not present in metadata, default value is data = N/2, parity = N/2
+func getDrivesCount(sc string, disks []StorageAPI) (data, parity int) {
+	totalDisks := len(disks)
+	switch sc {
+	case reducedRedundancyStorageClass:
+		if globalRRStorageClass.Parity != 0 {
+			// set the rrs parity if available
+			parity = globalRRStorageClass.Parity
+		} else {
+			// else fall back to default value
+			parity = defaultRRSParity
+		}
+	case standardStorageClass:
+		if globalStandardStorageClass.Parity != 0 {
+			// set the standard parity if available
+			parity = globalStandardStorageClass.Parity
+		} else {
+			// else fall back to default value
+			parity = totalDisks / 2
+		}
+	default:
+		parity = totalDisks / 2
+	}
+	// data is always totalDisks - parity
+	return totalDisks - parity, parity
+}
+
+// Returns per object readQuorum and writeQuorum
+// readQuorum is the minimum required disks to read data.
+// writeQuorum is the minimum required disks to write data.
+func objectQuorumFromMeta(xl xlObjects, partsMetaData []xlMetaV1, errs []error) (objectReadQuorum, objectWriteQuorum int, err error) {
+
+	// get the latest updated XL Metadata
+	latestXLMeta := listLatestXLMeta(partsMetaData, errs)
+
+	var validXLMeta xlMetaV1
+	emptyXLMetaV1 := xlMetaV1{}
+	count := 0
+
+	// latestXLMeta is a slice of xlMeta(s) updated most recently.
+	// We implicitly assume that all the xlMeta(s) have same dataBlocks and parityBlocks.
+	// Now loop through latestXLMeta and to get a count of valid xlMeta(s)
+	for _, xlMeta := range latestXLMeta {
+		if reflect.DeepEqual(xlMeta, emptyXLMetaV1) {
+			continue
```

Comment:
If you use continue you don't need the else block. 

```
if reflect.DeepEqual(xlMeta, emptyXLMetaV1) {
     continue
}
count++
validXLMeta = xlMeta
```

Reply:
Or negate condition of if statement:
```
if !reflect.DeepEqual(xlMeta, emptyXLMetaV1) {
    count++
    validXLMeta = xlMeta
}
```

Reply:
On a second though, if we let `listLastXLMeta` return a count then we can remove this loop and we also do not need the (expensive) `reflect.DeepEqual` call.

listLastXLMeta would end like this:
```
	count := 0
	for index, t := range modTimes {
 		if t == modTime {
 			latestXLMeta[index] = partsMetadata[index]
 			count++
 		}
 	}
 	// This is a slice with either partsMetaData that were updated the latest
 	// or empty partsMetaData
 	return latestXLMeta, count 
 }
```

Reply:
>  if we let listLastXLMeta return a count then we can remove this loop and we also do not need the (expensive) reflect.DeepEqual call.

Good idea, will do this.